core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java

```diff
@@ -261,31 +266,28 @@ private int updateShardStatuses(EnumSet<ShardStatusChangeType> changes, List<Ind
             while (statusShardIdIterator.hasNext()) {
                 ShardId statusShardId = statusShardIdIterator.next();
                 IndexService indexService = indicesService.indexService(statusShardId.getIndex());
-                boolean remove = false;
-                try {
-                    if (indexService == null) {
-                        remove = true;
-                        continue;
-                    }
+                boolean remove;
+                if (indexService == null) {
+                    remove = true;
+                } else {
                     IndexShard indexShard = indexService.shard(statusShardId.id());
                     if (indexShard == null) {
                         remove = true;
-                        continue;
-                    }
-                    remove = !CAN_UPDATE_INDEX_BUFFER_STATES.contains(indexShard.state());
-
-                } finally {
-                    if (remove) {
-                        changes.add(ShardStatusChangeType.DELETED);
-                        statusShardIdIterator.remove();
+                    } else {
+                        remove = !CAN_UPDATE_INDEX_BUFFER_STATES.contains(indexShard.state());
                     }
                 }
+                if (remove) {
+                    changes.add(ShardStatusChangeType.DELETED);
+                    statusShardIdIterator.remove();
+                }
             }
             return changes;
         }
 
         private void calcAndSetShardBuffers(int activeShards, String reason) {
             if (activeShards == 0) {
+                logger.debug("no active shards (reason={})", reason);
```

Comment:
I'd like to leave this one as debug: it should not run very frequently?  And there is already another debug line below for the more common case when there are active shards on this node ...